File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    int listLength(ListNode*head){
        int n = 0;
        while(head!=NULL){
            head = head->next;
            n++;
        }
        return n;
    }
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        int len = listLength(head);
        if(len==1) return NULL;
        if(n==len) return head->next;

        int pos = len-n;
        ListNode*curr = head;
        while(curr!=NULL and pos>1){
            curr = curr->next;
            pos--;
        }
        
        ListNode *next = curr->next->next;
        delete curr->next;
        curr->next = next;
        return head;

    }
};

File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        N, curr = 0, head
        while curr!=None:
            N+=1
            curr = curr.next
        
        n = N-n
        
        if n==0:
            return head.next
        
        curr = head

        while n>1:
            n-=1
            curr = curr.next
        
        curr.next = curr.next.next

        return head
